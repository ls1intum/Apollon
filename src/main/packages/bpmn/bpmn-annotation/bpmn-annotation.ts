import { BPMNElementType } from '..';
import { ILayer } from '../../../services/layouter/layer';
import { ILayoutable } from '../../../services/layouter/layoutable';
import { UMLElement } from '../../../services/uml-element/uml-element';
import { calculateNameBounds } from '../../../utils/name-bounds';
import { UMLElementType } from '../../uml-element-type';
import { UMLElementFeatures } from '../../../services/uml-element/uml-element-features';

export class BPMNAnnotation extends UMLElement {
  static features: UMLElementFeatures = { ...UMLElement.features };

  type: UMLElementType = BPMNElementType.BPMNAnnotation;

  render(canvas: ILayer): ILayoutable[] {
    this.bounds = calculateNameBounds(this, canvas);
    return [this];
  }
}
