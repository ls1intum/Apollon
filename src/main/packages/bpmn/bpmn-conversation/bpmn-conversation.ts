import { BPMNElementType } from '..';
import { ILayer } from '../../../services/layouter/layer';
import { ILayoutable } from '../../../services/layouter/layoutable';
import { IUMLElement, UMLElement } from '../../../services/uml-element/uml-element';
import { UMLElementType } from '../../uml-element-type';
import { UMLElementFeatures } from '../../../services/uml-element/uml-element-features';
import { IBoundary } from '../../../utils/geometry/boundary';
import { DeepPartial } from 'redux';
import { assign } from '../../../utils/fx/assign';
import * as Apollon from '../../../typings';

export type BPMNConversationType = 'default' | 'call';

export class BPMNConversation extends UMLElement {
  static features: UMLElementFeatures = { ...UMLElement.features, resizable: false };
  static defaultConversationType: BPMNConversationType = 'default';

  type: UMLElementType = BPMNElementType.BPMNConversation;
  bounds: IBoundary = { ...this.bounds, width: 40, height: 40 };
  conversationType: BPMNConversationType;

  constructor(values?: DeepPartial<BPMNConversation>) {
    super(values);
    assign<IUMLElement>(this, values);
    this.conversationType = values?.conversationType || BPMNConversation.defaultConversationType;
  }

  serialize(children?: UMLElement[]): Apollon.BPMNConversation {
    return {
      ...super.serialize(),
      type: this.type as keyof typeof BPMNElementType,
      conversationType: this.conversationType,
    };
  }

  deserialize<T extends Apollon.UMLModelElement>(
    values: T & { conversationType?: BPMNConversationType },
    children?: Apollon.UMLModelElement[],
  ): void {
    super.deserialize(values, children);
    this.conversationType = values.conversationType || BPMNConversation.defaultConversationType;
  }

  render(canvas: ILayer): ILayoutable[] {
    return [this];
  }
}
