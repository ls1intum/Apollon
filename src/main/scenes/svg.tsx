import React, { Component } from 'react';
import { DeepPartial } from 'redux';
import { defaults, Styles } from '../components/theme/styles';
import { Components } from '../packages/components';
import { UMLElementType } from '../packages/uml-element-type';
import { UMLElements } from '../packages/uml-elements';
import { UMLRelationshipType } from '../packages/uml-relationship-type';
import { UMLRelationships } from '../packages/uml-relationships';
import { ILayer } from '../services/layouter/layer';
import { UMLContainer } from '../services/uml-container/uml-container';
import { UMLElement } from '../services/uml-element/uml-element';
import { UMLRelationship } from '../services/uml-relationship/uml-relationship';
import * as Apollon from '../typings';
import { computeBoundingBoxForElements, IBoundary } from '../utils/geometry/boundary';
import { Point } from '../utils/geometry/point';
import { update } from '../utils/update';
import { Style } from './svg-styles';
import { StoreProvider } from '../components/store/model-store';
import { ModelState } from '../components/store/model-state';
import { ThemeProvider } from 'styled-components';

type Props = {
  model: Apollon.UMLModel;
  options?: Apollon.ExportOptions;
  styles?: DeepPartial<Styles>;
};

type State = {
  bounds: IBoundary;
  elements: UMLElement[];
};

const includeChildren = (
  elements: { [id: string]: UMLElement },
  ids: Set<string>,
  include: Set<string>,
): Set<string> => {
  const result = new Set<string>();
  for (const id of ids) {
    const element = elements[id];
    if (!element) continue;

    const children = new Set<string>(UMLContainer.isUMLContainer(element) ? element.ownedElements : []);
    if (include.has(id)) {
      result.add(id);
      include = new Set<string>([...include, ...children]);
    }
    includeChildren(elements, children, include).forEach(result.add, result);
  }
  return result;
};

const excludeChildren = (
  elements: { [id: string]: UMLElement },
  ids: Set<string>,
  exclude: Set<string>,
): Set<string> => {
  const result = new Set<string>();
  for (const id of ids) {
    const element = elements[id];
    if (!element) continue;

    const children = new Set<string>(UMLContainer.isUMLContainer(element) ? element.ownedElements : []);
    if (!exclude.has(id)) {
      result.add(id);
    } else {
      exclude = new Set<string>([...exclude, ...children]);
    }
    excludeChildren(elements, children, exclude).forEach(result.add, result);
  }
  return result;
};

const getInitialState = ({ model, options }: Props): State => {
  const layer: ILayer = {
    layer: document.createElementNS('http://www.w3.org/2000/svg', 'svg'),
    origin: (): Point => new Point(),
  };
  const apollonElements = model.elements;
  const apollonRelationships = model.relationships;

  const deserialize = (apollonElement: Apollon.UMLElement): UMLElement[] => {
    const element = new UMLElements[apollonElement.type]();
    const apollonChildren: Apollon.UMLElement[] = UMLContainer.isUMLContainer(element)
      ? apollonElements.filter((child) => child.owner === apollonElement.id)
      : [];

    element.deserialize(apollonElement, apollonChildren);
    const [root, ...updates] = element.render(layer, []) as UMLElement[];

    return [root, ...updates];
  };

  const elements = apollonElements.reduce<UMLElement[]>((acc, val) => [...acc, ...deserialize(val)], []);

  const relationships = apollonRelationships.map<UMLRelationship>((apollonRelationship) => {
    const relationship = new UMLRelationships[apollonRelationship.type]();
    relationship.deserialize(apollonRelationship);
    return relationship;
  });

  const elementState = [...elements, ...relationships].reduce<{ [id: string]: UMLElement }>(
    (acc, val) => ({ ...acc, [val.id]: val }),
    {},
  );
  const roots = Object.values(elementState).filter((element) => !element.owner);

  let layout = new Set<string>(Object.values(elementState).map((x) => x.id));
  if (options && options.include) {
    layout = includeChildren(
      elementState,
      new Set<string>(roots.map((element) => element.id)),
      new Set<string>(options.include),
    );
  }
  if (options && options.exclude) {
    layout = excludeChildren(
      elementState,
      new Set<string>(roots.map((element) => element.id)),
      new Set<string>(options.exclude),
    );
  }

  const keepOriginalSize = (options && options.keepOriginalSize) || false;

  const bounds = computeBoundingBoxForElements(
    Object.values(elementState).filter((element) => keepOriginalSize || layout.has(element.id)),
  );

  const margin = getMargin(options?.margin);
  bounds.x -= margin.left;
  bounds.y -= margin.top;
  bounds.width += margin.left + margin.right;
  bounds.height += margin.top + margin.bottom;

  const state = Object.values(elementState)
    .filter((element) => layout.has(element.id))
    .map((element) => {
      element.bounds.x -= bounds.x;
      element.bounds.y -= bounds.y;
      return element;
    });

  return {
    elements: state,
    bounds,
  };
};

const getMargin = (
  margin: Apollon.ExportOptions['margin'] = 15,
): { top: number; right: number; bottom: number; left: number } => {
  if (typeof margin === 'number') {
    return { top: margin, right: margin, bottom: margin, left: margin };
  }
  const result = { top: 0, right: 0, bottom: 0, left: 0 };
  return Object.assign(result, margin);
};

export class Svg extends Component<Props, State> {
  state = getInitialState(this.props);

  render() {
    const { bounds, elements } = this.state;
    const theme: Styles = update(defaults, this.props.styles || {});

    // connect exported svg to redux state, so that connected components can retrieve properties from state
    const state = ModelState.fromModel(this.props.model);

    return (
      <StoreProvider initialState={state}>
        <ThemeProvider theme={theme}>
          <svg
            width={bounds.width + 1}
            height={bounds.height + 1}
            xmlns="http://www.w3.org/2000/svg"
            xmlnsXlink="http://www.w3.org/1999/xlink"
            fillOpacity={0}
            fill="white"
          >
            <defs>
              <style>{(Style[0] as any)({ theme })}</style>
            </defs>
            {elements.map((element, index) => {
              const ElementComponent = Components[element.type as UMLElementType | UMLRelationshipType];
              return (
                <svg
                  {...element.bounds}
                  key={element.id}
                  className={element.name ? element.name.replace(/[<>]/, '') : ''}
                >
                  <ElementComponent key={index} element={element} />
                </svg>
              );
            })}
          </svg>
        </ThemeProvider>
      </StoreProvider>
    );
  }
}
