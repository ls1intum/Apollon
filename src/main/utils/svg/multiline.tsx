import React, { Component, CSSProperties, SVGProps } from 'react';
import { findDOMNode } from 'react-dom';

const defaultProps = Object.freeze({
  x: 0,
  y: 0,
  dx: 0,
  dy: 0,
  angle: 0,
  width: undefined as number | undefined,
  height: undefined as number | undefined,
  lineHeight: 16,
  capHeight: 11,
  scaleToFit: false,
  textAnchor: 'middle' as 'start' | 'middle' | 'end' | 'inherit',
  verticalAnchor: 'middle' as 'start' | 'middle' | 'end',
});

const getInitialState = (props: Props) => {
  const words = props.children ? props.children.toString().split(/\s+/) : [];
  return { wordsByLines: [{ words, width: 0 }] };
};

type Props = { children: string } & SVGProps<SVGTextElement> & typeof defaultProps;

type State = typeof getInitialState;

export class Multiline extends Component<Props, State> {
  static defaultProps = defaultProps;
  state = getInitialState(this.props);

  spaceWidth = 0;
  wordsWithComputedWidth = [] as { word: string; width: number }[];

  componentDidMount() {
    this.updateWordsByLines(this.props, true);
  }

  componentDidUpdate(previousProps: Readonly<Props>) {
    const needCalculate = previousProps.children !== this.props.children || previousProps.style !== this.props.style;
    if (needCalculate) {
      this.updateWordsByLines(this.props, needCalculate);
    }
  }

  calculateWordWidths = (props: Readonly<Props>) => {
    try {
      const words = props.children ? props.children.toString().split(/\s+/) : [];
      const wordsWithComputedWidth = words.map((word) => ({ word, width: this.getStringWidth(word, props.style) }));

      const spaceWidth = this.getStringWidth('\u00A0', props.style);

      return { wordsWithComputedWidth, spaceWidth };
    } catch (e) {
      return null;
    }
  };

  getStringWidth(str: string, style?: CSSProperties): number {
    try {
      // Calculate length of each word to be used to determine number of words per line
      const container = findDOMNode(this);
      if (!container) {
        return 0;
      }

      const text = document.createElementNS('http://www.w3.org/2000/svg', 'tspan');
      container.appendChild(text);

      Object.assign(text.style, style);
      text.textContent = str;
      const width = text.getComputedTextLength();

      container.removeChild(text);

      return width;
    } catch (e) {
      return 0;
    }
  }

  updateWordsByLines(props: Readonly<Props>, needCalculate: boolean) {
    // Only perform calculations if using features that require them (multiline, scaleToFit)
    if (props.width || props.scaleToFit) {
      if (needCalculate) {
        const wordWidths = this.calculateWordWidths(props);

        if (wordWidths) {
          const { wordsWithComputedWidth, spaceWidth } = wordWidths;

          this.wordsWithComputedWidth = wordsWithComputedWidth;
          this.spaceWidth = spaceWidth;
        } else {
          this.updateWordsWithoutCalculate(props);

          return;
        }
      }

      const wordsByLines = this.calculateWordsByLines(this.wordsWithComputedWidth, this.spaceWidth, props.width);
      this.setState({ wordsByLines });
    } else {
      this.updateWordsWithoutCalculate(props);
    }
  }

  updateWordsWithoutCalculate(props: Readonly<Props>) {
    const words = props.children ? props.children.toString().split(/\s+/) : [];
    this.setState({ wordsByLines: [{ words, width: 0 }] });
  }

  calculateWordsByLines(
    wordsWithComputedWidth: { word: string; width: number }[],
    spaceWidth: number,
    lineWidth?: number,
  ): { words: string[]; width: number }[] {
    const { scaleToFit } = this.props;
    return wordsWithComputedWidth.reduce<{ words: string[]; width: number }[]>((result, { word, width }) => {
      const currentLine = result[result.length - 1];

      if (currentLine && (!lineWidth || scaleToFit || currentLine.width + width + spaceWidth < lineWidth)) {
        // Word can be added to an existing line
        currentLine.words.push(word);
        currentLine.width += width + spaceWidth;
      } else {
        // Add first word to line or word is too long to scaleToFit on existing line
        const newLine = { words: [word], width };
        result.push(newLine);
      }

      return result;
    }, []);
  }

  render() {
    const { x, y, dx, dy, textAnchor, verticalAnchor, scaleToFit, angle, lineHeight, capHeight, fill, ...textProps } =
      this.props;
    const { wordsByLines } = this.state;

    const xPosition = x + dx;
    const yPosition = y + dy;

    let startDy: string | number | undefined;
    switch (verticalAnchor) {
      case 'start':
        startDy = capHeight;
        break;
      case 'middle':
        startDy = ((wordsByLines.length - 1) / 2) * -lineHeight + capHeight / 2;
        break;
      default:
        startDy = wordsByLines.length - 1 * -lineHeight;
        break;
    }

    const transforms = [];
    if (scaleToFit && wordsByLines.length) {
      const lineWidth = wordsByLines[0].width;
      const sx = (this.props.width || 0) / lineWidth;
      const sy = sx;
      const originX = xPosition - sx * xPosition;
      const originY = yPosition - sy * yPosition;
      transforms.push(`matrix(${sx}, 0, 0, ${sy}, ${originX}, ${originY})`);
    }
    if (angle) {
      transforms.push(`rotate(${angle}, ${xPosition}, ${yPosition})`);
    }
    if (transforms.length) {
      textProps.transform = transforms.join(' ');
    }

    return (
      <text
        style={fill ? { fill } : {}}
        x={xPosition}
        y={yPosition}
        textAnchor={textAnchor}
        {...textProps}
        pointerEvents="none"
      >
        {wordsByLines.map((line, index) => (
          <tspan x={xPosition} dy={index === 0 ? startDy : lineHeight} key={index}>
            {line.words.join(' ')}
          </tspan>
        ))}
      </text>
    );
  }
}
