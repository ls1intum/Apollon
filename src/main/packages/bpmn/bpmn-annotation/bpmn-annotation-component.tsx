import React, { ComponentType, FunctionComponent } from 'react';
import { ThemedPath, ThemedRect } from '../../../components/theme/themedComponents';
import { Multiline } from '../../../utils/svg/multiline';
import { BPMNGroup } from '../bpmn-group/bpmn-group';
import { withTheme, withThemeProps } from '../../../components/theme/styles';
import { compose } from 'redux';
import { connect, ConnectedComponent } from 'react-redux';
import { ModelState } from '../../../components/store/model-state';
import { ApollonView } from '../../../services/editor/editor-types';

type OwnProps = {
  element: BPMNGroup;
  strokeColor?: string;
  textColor?: string;
  children?: React.ReactNode;
};

type StateProps = { interactive: boolean; interactable: boolean; hovered: boolean };

type DispatchProps = {};

type Props = OwnProps & StateProps & DispatchProps & withThemeProps;

const enhance = compose<ConnectedComponent<ComponentType<Props>, OwnProps>>(
  withTheme,
  connect<StateProps, DispatchProps, OwnProps, ModelState>((state, props) => ({
    hovered: state.hovered.includes(props.element.id),
    interactive: state.interactive.includes(props.element.id),
    interactable: state.editor.view === ApollonView.Exporting || state.editor.view === ApollonView.Highlight,
  })),
);

export const BPMNAnnotationC: FunctionComponent<Props> = ({
  element,
  strokeColor,
  textColor,
  interactive,
  interactable,
  hovered,
  theme,
}) => (
  <g>
    <ThemedRect
      rx={10}
      ry={10}
      width={element.bounds.width}
      height={element.bounds.height}
      strokeColor="transparent"
      fillColor={
        interactable && interactive
          ? theme.interactive.normal
          : interactable && hovered
            ? theme.interactive.hovered
            : 'transparent'
      }
    />
    <ThemedPath
      d={`M20,0 L10,0 A 10 10 280 0 0 0 10 L0,${element.bounds.height - 10} A 10 10 180 0 0 10 ${
        element.bounds.height
      } L20, ${element.bounds.height}`}
      strokeColor={strokeColor || element.strokeColor}
      fillColor="transparent"
    />
    <Multiline
      x={element.bounds.width / 2}
      y={element.bounds.height / 2}
      width={element.bounds.width}
      height={element.bounds.height}
      fontWeight="bold"
      fill={textColor || element.textColor}
      lineHeight={16}
      capHeight={11}
    >
      {element.name}
    </Multiline>
  </g>
);

export const BPMNAnnotationComponent = enhance(BPMNAnnotationC);
