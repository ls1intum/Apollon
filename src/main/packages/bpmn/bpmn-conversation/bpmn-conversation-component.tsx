import React, { FunctionComponent } from 'react';
import { BPMNConversation } from './bpmn-conversation';
import { ThemedPolyline } from '../../../components/theme/themedComponents';
import { Multiline } from '../../../utils/svg/multiline';

export const BPMNConversationComponent: FunctionComponent<Props> = ({ element }) => (
  <g>
    <ThemedPolyline
      points={`${element.bounds.width / 4} 0, ${(element.bounds.width / 4) * 3} 0, ${element.bounds.width} ${
        element.bounds.height / 2
      }, ${(element.bounds.width / 4) * 3} ${element.bounds.height}, ${element.bounds.width / 4} ${
        element.bounds.height
      }, 0 ${element.bounds.height / 2}, ${element.bounds.width / 4} 0`}
      strokeColor={element.strokeColor}
      fillColor={element.fillColor}
      strokeWidth={element.conversationType === 'call' ? 3 : 1}
    />
    <Multiline
      x={element.bounds.width + 10}
      y={element.bounds.height / 2}
      width={element.bounds.width}
      height={element.bounds.height}
      fill={element.textColor}
      lineHeight={16}
      capHeight={11}
      textAnchor="start"
    >
      {element.name}
    </Multiline>
  </g>
);

export interface Props {
  element: BPMNConversation;
}
