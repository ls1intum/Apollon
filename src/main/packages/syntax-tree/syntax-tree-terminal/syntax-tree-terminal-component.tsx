import React, { FunctionComponent } from 'react';
import { Multiline } from '../../../utils/svg/multiline';
import { SyntaxTreeTerminal } from './syntax-tree-terminal';
import { ThemedRect } from '../../../components/theme/themedComponents';

export const SyntaxTreeTerminalComponent: FunctionComponent<Props> = ({ element, scale }) => (
  <g>
    <ThemedRect width="100%" height="100%" strokeColor={element.strokeColor} />
    <Multiline
      x={element.bounds.width / 2}
      y={element.bounds.height / 2}
      width={element.bounds.width}
      height={element.bounds.height}
      fontWeight="bold"
      fill={element.textColor}
      lineHeight={16 * scale}
      capHeight={11 * scale}
    >
      {element.name}
    </Multiline>
  </g>
);

export interface Props {
  element: SyntaxTreeTerminal;
  scale: number;
}
