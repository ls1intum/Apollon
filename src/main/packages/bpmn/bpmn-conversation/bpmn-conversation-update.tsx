import React, { Component, ComponentClass } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { Button } from '../../../components/controls/button/button';
import { Divider } from '../../../components/controls/divider/divider';
import { TrashIcon } from '../../../components/controls/icon/trash';
import { Textfield } from '../../../components/controls/textfield/textfield';
import { I18nContext } from '../../../components/i18n/i18n-context';
import { localized } from '../../../components/i18n/localized';
import { ModelState } from '../../../components/store/model-state';
import { styled } from '../../../components/theme/styles';
import { UMLElementRepository } from '../../../services/uml-element/uml-element-repository';
import { Switch } from '../../../components/controls/switch/switch';
import { BPMNConversation, BPMNConversationType } from './bpmn-conversation';
import { Dropdown } from '../../../components/controls/dropdown/dropdown';

interface OwnProps {
  element: BPMNConversation;
}

type StateProps = {};

interface DispatchProps {
  update: typeof UMLElementRepository.update;
  delete: typeof UMLElementRepository.delete;
}

type Props = OwnProps & StateProps & DispatchProps & I18nContext;

const enhance = compose<ComponentClass<OwnProps>>(
  localized,
  connect<StateProps, DispatchProps, OwnProps, ModelState>(null, {
    update: UMLElementRepository.update,
    delete: UMLElementRepository.delete,
  }),
);

const Flex = styled.div`
  display: flex;
  align-items: baseline;
  justify-content: space-between;
`;

class BPMNConversationUpdateComponent extends Component<Props> {
  render() {
    const { element } = this.props;

    return (
      <div>
        <section>
          <Flex>
            <Textfield value={element.name} onChange={this.rename(element.id)} autoFocus />
            <Button color="link" tabIndex={-1} onClick={this.delete(element.id)}>
              <TrashIcon />
            </Button>
          </Flex>
          <Divider />
        </section>
        <section>
          <Dropdown value={element.conversationType} onChange={this.changeConversationType(element.id)} color="primary">
            <Dropdown.Item value={'default'}>{this.props.translate('packages.BPMN.BPMNConversation')}</Dropdown.Item>
            <Switch.Item value={'call'}>{this.props.translate('packages.BPMN.BPMNCallConversation')}</Switch.Item>
          </Dropdown>
        </section>
      </div>
    );
  }

  /**
   * Rename the conversation
   * @param id The ID of the conversation that should be renamed
   */
  private rename = (id: string) => (value: string) => {
    this.props.update(id, { name: value });
  };

  /**
   * Change the type of the conversation
   * @param id The ID of the conversation whose type should be changed
   */
  private changeConversationType = (id: string) => (value: string) => {
    this.props.update<BPMNConversation>(id, { conversationType: value as BPMNConversationType });
  };

  /**
   * Delete a conversation
   * @param id The ID of the conversation that should be deleted
   */
  private delete = (id: string) => () => {
    this.props.delete(id);
  };
}

export const BPMNConversationUpdate = enhance(BPMNConversationUpdateComponent);
