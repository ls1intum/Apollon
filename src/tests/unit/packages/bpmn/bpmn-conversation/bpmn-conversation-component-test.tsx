import { wrappedRender } from '../../../test-utils/render';
import * as React from 'react';
import { SyntaxTreeTerminal } from '../../../../../main/packages/syntax-tree/syntax-tree-terminal/syntax-tree-terminal';
import { SyntaxTreeTerminalComponent } from '../../../../../main/packages/syntax-tree/syntax-tree-terminal/syntax-tree-terminal-component';
import { Multiline } from '../../../../../main/utils/svg/multiline';
import { CSSProperties } from 'react';
import { BPMNAnnotation } from '../../../../../main/packages/bpmn/bpmn-annotation/bpmn-annotation';
import { BPMNAnnotationComponent } from '../../../../../main/packages/bpmn/bpmn-annotation/bpmn-annotation-component';
import { BPMNConversationComponent } from '../../../../../main/packages/bpmn/bpmn-conversation/bpmn-conversation-component';
import { BPMNConversation } from '../../../../../main/packages/bpmn/bpmn-conversation/bpmn-conversation';

// override getStringWidth, because it uses by jsdom unsupported SVGElement methods
Multiline.prototype.getStringWidth = (str: string, style?: CSSProperties) => {
  return 0;
};

it('render the bpmn-conversation-component', () => {
  const conversation: BPMNConversation = new BPMNConversation({ name: 'Conversation' });
  const { getByText, baseElement } = wrappedRender(
    <svg>
      <BPMNConversationComponent element={conversation} />
    </svg>,
  );
  expect(getByText(conversation.name)).toBeInTheDocument();
  expect(baseElement).toMatchSnapshot();
});
