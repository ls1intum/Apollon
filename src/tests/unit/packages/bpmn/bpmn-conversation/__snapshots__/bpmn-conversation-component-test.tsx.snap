// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the bpmn-conversation-component 1`] = `
<body>
  <div>
    <svg>
      <g>
        <polyline
          class="sc-aXZVg bUvEJV"
          fill="white"
          points="10 0, 30 0, 40 20, 30 40, 10 40, 0 20, 10 0"
          stroke="black"
          stroke-width="1"
        />
        <text
          height="40"
          pointer-events="none"
          text-anchor="start"
          width="40"
          x="50"
          y="20"
        >
          <tspan
            dy="5.5"
            x="50"
          >
            Conversation
          </tspan>
        </text>
      </g>
    </svg>
  </div>
</body>
`;
